### DISCUSSION TITLE
Dynamic refs not working

### DISCUSSION THREAD

**Discussion by mnabialek** [alpinejs/alpine#1975] [7. Bugs (archived)] [answered]
It seems it was working fine in v2 but there is problem in v3.

I have component like this:

```
'use strict';

export default () => ({
    options: [
        {value: 1},
        {value: 2},
        {value: 3}
    ],

   display()
   {
       console.log(this.$refs[0], this.$refs[1]);
   }
});
```

I start Alpine like this:

```
import Alpine from 'alpinejs';
window.Alpine = Alpine;

import Sample from "./modules/Sample";
window.Sample = Sample;

Alpine.data('Sample', Sample);
```

and now HTML code:

```
 <div x-data="Sample">
      <div x-ref="0" @mouseover="console.log($el); console.log($refs[0]);">0</div>
      <template x-for="option in options">
          <div :x-ref="option.value" x-text="option.value" @mouseover="console.log($el); console.log($el.getAttribute('x-ref')); console.log($refs[0], $refs[1]);" @click="display()"></div>
      </template>
  </div>
```

Result is that $refs[1] is not working inside component and on mouseover but ref created manually (outside of template) seems to be working without any problem

**Comment by SimoTod** [comment, marked as answer]
You can't bind x-ref using Alpine itself in v3 (:x-ref doesn't work anymore). See https://github.com/alpinejs/alpine/discussions/1527#discussioncomment-996994

**Comment by mnabialek** [comment]
Thanks, I've just noticed it a while ago. Unfortunately upgrade guide is missing that. I've solved it by using

```
x-bind:id="'some_prefix' + option.value"
```
and then just using

```
document.getElementById('some_prefix' + option.value)
```

**Reply by SimoTod** [reply]
Yeah, I think that confused a lot of people. The docs is actually in this repo now: packages/docs/src/en. If you send a PR, I'm sure Caleb will be happy to merge it in.

**Reply by mnabialek** [reply]
Thanks, I've just created it - https://github.com/alpinejs/alpine/pull/1976 Let's see what will happen

**Reply by rizkhal** [reply]
I working on form with clone function using `"alpinejs": "^3.4.2"` but this not working for me..

```js
<div x-data="{
  fields: [],
  init() {
   this.clone();
  },
  clone() {
     this.fields.push({ product: null });
  },
">
   <template x-for="(form, index) in fields" :key="index">
    <div x-data x-init="() => { let element = document.getElementById(`product-${index}`); }">
      <select data-type="select-one" :id="'product-' + index"></select>
    </div>
  </template>
  <button @click="clone()">Clone</button>
</div>
```

The element always `null` but the unique `id` is generated on brwoser like below
```blade
<select data-type="select-one" :id="'product-' + index" id="product-0"></select>
```

**Reply by SimoTod** [reply]
@rizkhal elements are evaluated top-down. When Alpine evaluates x-init, it hasn't evaluated :id="'product-' + index" yet.
You can either move x-init to the select tag (you don't need to move x-data if you don't have to) and it would work or wrap the x-init callback in `$nextTick` (`x-init="$nextTick(() => {...})")

**Reply by rizkhal** [reply]
Its working now, Thanks @SimoTod

**Reply by marzban-dev** [reply]
its working...
niceeeeeeeeeeeeeeeee👌
